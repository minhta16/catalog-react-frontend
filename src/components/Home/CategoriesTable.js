import React, { Component } from 'react';
import PropTypes from 'prop-types';
import randomMC from 'random-material-color';
import {
  Paper,
  Grid,
  MenuList,
  MenuItem,
  ListItem,
  List,
  Typography,
  Divider,
  CircularProgress,
  ListItemText,
  ListItemAvatar,
  Avatar,
} from '@material-ui/core';

import { Link } from 'react-router-dom';

class CategoriesTable extends Component {
  state = {};

  getShortDescription = (description) => {
    let retString = description.substr(0, 100);
    retString += description.length > 100 ? '...' : '';
    return retString;
  };

  getSecondaryText = (item) => (
    <React.Fragment>
      <Typography component="span" variant="body2" style={{ display: 'inline' }}>
        {`$${Number(item.price).toFixed(2)} - `}
      </Typography>
      {this.getShortDescription(item.description)}
    </React.Fragment>
  );

  render() {
    const {
      categories,
      selectedCatId,
      selectedCatItems,
      fetchPosts,
      selectedCat,
      categoriesLoading,
      postsLoading,
    } = this.props;
    const selectedCatName = selectedCat ? selectedCat.name : '';
    return (
      <Paper>
        {/* Create a loading animation if the categories are loading */}
        {categoriesLoading ? (
          <CircularProgress />
        ) : (
          <Grid container spacing={2}>
            <Grid item xs={3} className="vertical-divider">
              <Grid className="mainmenu-categories-header">
                <Typography variant="h4">Categories</Typography>
              </Grid>
              <MenuList>
                {categories.map((category) => (
                  <MenuItem
                    className="categoryMenuItem"
                    key={category.id}
                    selected={String(selectedCatId) === String(category.id)}
                    component={Link}
                    to={`/category/${category.id}`}
                    onClick={() => fetchPosts(category.id)}
                  >
                    {category.name}
                  </MenuItem>
                ))}
              </MenuList>
            </Grid>
            <Grid item xs={9}>
              {/* Create a loading animation if the posts are loading */}
              {postsLoading ? (
                <CircularProgress />
              ) : (
                <div>
                  <Typography id="categories-table-name-typo" className="left-margin" variant="h4">
                    {selectedCatId ? selectedCatName : 'Choose a Category'}
                  </Typography>
                  <Divider variant="middle" />
                  {/* A list of items to be selected, click to redirect */}
                  <List>
                    {/* If selectedCatID is available then display the menu. If not, the right side will display a message telling the user to select a category */}
                    {selectedCatId ? (
                      selectedCatItems.map((item) => (
                        <ListItem
                          button
                          className="itemsMenuItem"
                          key={item.id}
                          component={Link}
                          to={`/category/${selectedCatId}/post/${item.id}`}
                        >
                          <ListItemAvatar>
                            <Avatar
                              style={{
                                backgroundColor: randomMC.getColor({
                                  shades: ['300'],
                                  text: item.name,
                                }),
                                color: '#fff',
                              }}
                            >
                              {item.name.charAt(0)}
                            </Avatar>
                          </ListItemAvatar>
                          <ListItemText
                            primary={item.name}
                            secondary={this.getSecondaryText(item)}
                          />
                        </ListItem>
                      ))
                    ) : (
                      <Typography className="itemTypography" variant="body1">
                        Please select a category on the left to view its items.
                      </Typography>
                    )}
                  </List>
                </div>
              )}
            </Grid>
          </Grid>
        )}
      </Paper>
    );
  }
}

CategoriesTable.propTypes = {
  selectedCatId: PropTypes.string,
  categories: PropTypes.array.isRequired,
  selectedCatItems: PropTypes.array.isRequired,
  fetchPosts: PropTypes.func.isRequired,
  selectedCat: PropTypes.object.isRequired,
  categoriesLoading: PropTypes.bool.isRequired,
  postsLoading: PropTypes.bool.isRequired,
};

CategoriesTable.defaultProps = {
  selectedCatId: '',
};

export default CategoriesTable;
