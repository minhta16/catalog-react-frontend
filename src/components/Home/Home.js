import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Container } from '@material-ui/core';
import CategoriesTable from 'components/Home/CategoriesTable';
import { fetchPosts as fetchPostsRedux } from 'actions/posts';
import { selectAllPosts, selectPostLoading } from 'selectors/posts';
import { selectCategories, selectCategory, selectCategoriesLoading } from 'selectors/categories';
import InfoSnackbar from 'components/Shared/InfoSnackbar';

export class Home extends Component {
  state = {
    openSnackbar: false,
  };

  /**
   * always fetchCategories upon mount. if the user is viewing a category then fetch the posts in that category. If there is a snackbarMessage then open the snackbar
   */
  componentDidMount() {
    const { match, fetchPosts, location } = this.props;
    if (match.params.id) {
      fetchPosts(match.params.id);
    }
    if (location.snackbarMess) {
      this.closeSnackbar(false)();
    }
  }

  closeSnackbar = (close) => () => {
    this.setState({
      openSnackbar: !close,
    });
  };

  render() {
    const {
      match,
      categories,
      selectedCatItems,
      fetchPosts,
      selectedCategory,
      location,
      categoriesLoading,
      postsLoading,
    } = this.props;
    const { openSnackbar } = this.state;
    // Render a categories table and a snackbar to be displayed when needed
    return (
      <Container maxWidth="lg">
        <CategoriesTable
          categories={categories}
          selectedCatId={match.params.id}
          selectedCat={selectedCategory}
          selectedCatItems={selectedCatItems}
          fetchPosts={fetchPosts}
          categoriesLoading={categoriesLoading}
          postsLoading={postsLoading}
        />

        <InfoSnackbar
          open={openSnackbar}
          onClose={this.closeSnackbar(true)}
          message={location.snackbarMess}
        />
      </Container>
    );
  }
}

Home.propTypes = {
  match: PropTypes.object,
  categories: PropTypes.array,
  fetchPosts: PropTypes.func.isRequired,
  selectedCatItems: PropTypes.array,
  selectedCategory: PropTypes.object,
  location: PropTypes.object.isRequired,
  categoriesLoading: PropTypes.bool.isRequired,
  postsLoading: PropTypes.bool.isRequired,
};

Home.defaultProps = {
  match: {},
  categories: {},
  selectedCatItems: {},
  selectedCategory: {},
};

const mapSelectorToProps = (state, ownProps) => ({
  selectedCatItems: selectAllPosts(state, 'reverse'),
  categories: selectCategories(state),
  selectedCategory: selectCategory(state, ownProps.match.params.id),
  categoriesLoading: selectCategoriesLoading(state),
  postsLoading: selectPostLoading(state),
});

const mapDispatchToProps = {
  fetchPosts: fetchPostsRedux,
};

export default connect(
  mapSelectorToProps,
  mapDispatchToProps,
)(Home);
