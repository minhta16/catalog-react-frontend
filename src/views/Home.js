import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Container } from '@material-ui/core';
import CategoriesTable from 'components/CategoriesTable';
import 'App.css';

import { fetchCategories as fetchCategoriesRedux } from 'references/redux/actions/categories';
import { fetchPosts as fetchPostsRedux } from 'references/redux/actions/posts';
import { selectAllPosts } from 'references/redux/selectors/posts';
import { selectCategories, selectCategory } from 'references/redux/selectors/categories';

export class Home extends Component {
  componentDidMount() {
    const { fetchCategories, match, fetchPosts } = this.props;
    fetchCategories();
    if (match.params.id) {
      fetchPosts(match.params.id);
    }
  }

  render() {
    const { match, categories, selectedCatItems, fetchPosts, selectedCategory } = this.props;
    return (
      <Container maxWidth="lg">
        <CategoriesTable
          categories={categories}
          selectedCatId={match.params.id}
          selectedCat={selectedCategory}
          selectedCatItems={selectedCatItems}
          fetchPosts={fetchPosts}
        />
      </Container>
    );
  }
}

Home.propTypes = {
  match: PropTypes.object,
  categories: PropTypes.array,
  fetchCategories: PropTypes.func.isRequired,
  fetchPosts: PropTypes.func.isRequired,
  selectedCatItems: PropTypes.array,
  selectedCategory: PropTypes.object,
};

Home.defaultProps = {
  match: {},
  categories: {},
  selectedCatItems: {},
  selectedCategory: {},
};

const mapSelectorToProps = (state, ownProps) => ({
  selectedCatItems: selectAllPosts(state),
  categories: selectCategories(state),
  selectedCategory: selectCategory(state, ownProps.match.params.id),
});

const mapDispatchToProps = {
  fetchCategories: fetchCategoriesRedux,
  fetchPosts: fetchPostsRedux,
};

export default connect(
  mapSelectorToProps,
  mapDispatchToProps,
)(Home);
