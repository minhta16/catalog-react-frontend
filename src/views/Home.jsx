import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Container } from '@material-ui/core';
import CategoriesTable from 'components/CategoriesTable';
import 'App.css';

class Home extends Component {
  componentDidMount() {
    return 0;
  }

  render() {
    return (
      <Container maxWidth="lg">
        <CategoriesTable />
      </Container>
    );
  }
}

const mapStateToProps = state => ({
  categories: state.categories,
});

export default connect(mapStateToProps)(Home);
