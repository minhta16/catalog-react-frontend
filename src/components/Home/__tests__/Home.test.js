/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { Home } from '../Home';

describe('components/Home/Home', () => {
  let wrapper;
  let props;
  // let warning;

  // const update = () => {
  //   wrapper.update();
  //   wrapper.instance().forceUpdate();
  // };

  const setup = () => {
    wrapper = shallow(<Home {...props} />);
  };

  beforeEach(() => {
    props = {
      fetchCategories: jest.fn(),
      match: {
        params: {
          id: '1',
        },
      },
      location: {
        snackbarMess: 'true',
      },
      fetchPosts: jest.fn(),
      categories: [
        {
          created: '2019',
          updated: '2019',
          description: 'testDesc',
          name: 'testName',
          id: '1',
        },
      ],
      selectedCatItems: [],
      categoriesLoading: false,
      postsLoading: false,
      openSnackbar: jest.fn(),
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('When view mounts', () => {
    it('should not call fetchPosts if id is blank', () => {
      props.match.params.id = '';
      setup();
      expect(props.fetchPosts).not.toHaveBeenCalled();
    });

    it('should call fetchPosts if id is something', () => {
      setup();
      expect(props.fetchPosts).toHaveBeenCalled();
    });
  });
});
