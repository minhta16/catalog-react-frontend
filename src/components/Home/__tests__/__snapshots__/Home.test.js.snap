// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Home/Home should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="lg"
>
  <CategoriesTable
    categories={
      Array [
        Object {
          "created": "2019",
          "description": "testDesc",
          "id": "1",
          "name": "testName",
          "updated": "2019",
        },
      ]
    }
    categoriesLoading={false}
    fetchPosts={
      [MockFunction] {
        "calls": Array [
          Array [
            "1",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    postsLoading={false}
    selectedCat={Object {}}
    selectedCatId="1"
    selectedCatItems={Array []}
  />
</WithStyles(ForwardRef(Container))>
`;
